src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,179 @@
+                newNode.getPrev().setNext(newNode);
+            }
+            node.setPrev(newNode);
+            if (index == 0) {
+                head = newNode;
+            }
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
+    }
+
+    private ListNode<T> getListNode(int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException(INDEX_OUT_OF_BOUNDS);
+        }
+
+        if (size == 0 || index == size) {
+            return null;
+        }
+
+        if (index == 0) {
+            return head;
+        }
+
+        if (index == size - 1) {
+            return tail;
+        }
+
+        ListNode<T> nextNode = head;
+        for (int i = 1; i <= index; i++) {
+            nextNode = nextNode.getNext();
+        }
+        return nextNode;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        ListNode<T> node = getListNode(index);
+        if (node == null) {
+            throw new IndexOutOfBoundsException(INDEX_OUT_OF_BOUNDS);
+        }
+        return node.getValue();
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        ListNode<T> node = getListNode(index);
+        if (node == null) {
+            throw new IndexOutOfBoundsException(INDEX_OUT_OF_BOUNDS);
+        }
+        T oldValue = node.getValue();
+        node.setValue(value);
+        return oldValue;
+    }
+
+    private void removeNode(ListNode<T> node) {
+        ListNode<T> nextNode = node.getNext();
+        ListNode<T> prevNode = node.getPrev();
+
+        if (nextNode != null) {
+            nextNode.setPrev(prevNode);
+        } else {
+            tail = size == 1 ? head : prevNode;
+        }
+        if (prevNode != null) {
+            prevNode.setNext(nextNode);
+        } else {
+            head = nextNode;
+        }
+        size--;
+        if (isEmpty()) {
+            head = null;
+            tail = null;
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        ListNode<T> node = getListNode(index);
+        if (node == null) {
+            throw new IndexOutOfBoundsException(INDEX_OUT_OF_BOUNDS);
+        }
+        removeNode(node);
+        return node.getValue();
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        ListNode<T> nextNode = head;
+
+        if (nextNode == null) {
+            return false;
+        }
+
+        while (!(nextNode.getValue() == object
+                || nextNode.getValue() != null && nextNode.getValue().equals(object))) {
+            nextNode = nextNode.getNext();
+            if (nextNode == null) {
+                break;
+            }
+        }
+        if (nextNode == null) {
+            return false;
+        }
+        removeNode(nextNode);
+        return true;
```

Comment:
Create a separate method unlink(Node node).
This logic will be used in 2 methods: when we remove from list by index and by value. So let’s create a method that will take a Node that needs to be removed and unlink it. As a result we will call it from both methods when Node is found.